datafusion/optimizer/src/eliminate_nested_union.rs

```diff
@@ -60,7 +60,8 @@ impl OptimizerRule for EliminateNestedUnion {
                 let inputs = inputs
                     .into_iter()
                     .flat_map(extract_plans_from_union)
-                    .collect::<Vec<_>>();
+                    .map(|plan| coerce_plan_expr_for_schema(&plan, &schema))
```

Comment:
This makes sense -- that the children were not all coerced with respect to the outermost union schema but rather with respect to the inner one.